File: Queue/649. Dota2 Senate/solution.ts
function predictPartyVictory(senate: string): string {
    const n = senate.length;
    const radiant: number[] = [];
    const dire: number[] = [];

    for (let i = 0; i < n; i++) {
        if (senate[i] === 'R') {
            radiant.push(i);
        } else {
            dire.push(i);
        }
    }

    while (radiant.length > 0 && dire.length > 0) {
        const radIdx = radiant.shift()!;
        const dirIdx = dire.shift()!;

        if (radIdx < dirIdx) {
            radiant.push(radIdx + n);
        } else {
            dire.push(dirIdx + n);
        }
    }

    return radiant.length > 0 ? "Radiant" : "Dire";
}

File: Queue/649. Dota2 Senate/solution.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        radiant, dire = deque(), deque()
        
        for i, party in enumerate(senate):
            if party == 'R':
                radiant.append(i)
            else:
                dire.append(i)
        
        while radiant and dire:
            rad_idx, dir_idx = radiant.popleft(), dire.popleft()
            
            if rad_idx < dir_idx:
                radiant.append(rad_idx + n)
            else:
                dire.append(dir_idx + n)
        
        return "Radiant" if radiant else "Dire"